File: Classes/Views/ChartView.m
//
//  ChartView.m
//  ShralpTide
//
//  Created by Michael Parlee on 9/22/08.
//  Copyright 2009 Michael Parlee. All rights reserved.
/*
   This file is part of ShralpTide.

   ShralpTide is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   ShralpTide is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with ShralpTide.  If not, see <http://www.gnu.org/licenses/>.
*/

#import "ChartView.h"
#import <QuartzCore/QuartzCore.h>
#import "NSDate+Day.h"

@interface ChartView ()
- (float)findLowestTide:(SDTide *)tide;
- (float)findHighestTide:(SDTide *)tide;
- (void)hideTideDetails;
- (NSDate*)midnight;
- (NSDate*)midnight:(NSDate*)date;

@property (assign) float xratio;
@property (assign) float yratio;
@property (assign) float yoffset;
@end

@implementation ChartView

- (id)initWithCoder:(NSCoder *)coder {
	if ((self = [super initWithCoder:coder])) {
        self.height = 234; // legacy default
        self.hoursToPlot = 24;
    }
    return self;
}

- (BOOL)isOpaque
{
    return YES;
}

- (NSDate*)endTime
{
    return [NSDate dateWithTimeIntervalSince1970:[[self.datasource day] timeIntervalSince1970] + (self.hoursToPlot * 60 * 60)];
}

- (NSArray*)pairRiseAndSetEvents:(NSArray*)events riseEventType:(SDTideState)riseType setEventType:(SDTideState)setType
{
    NSMutableArray *pairs = [[NSMutableArray alloc] init];
    NSDate *riseTime = nil;
    NSDate *setTime = nil;
    for (SDTideEvent *event in events) {
        if (event.eventType == riseType) {
            riseTime = event.eventTime;
            if ([event isEqual:[events lastObject]]) {
                setTime = self.endTime;
            }
        }
        if (event.eventType == setType) {
            if ([events indexOfObject:event] == 0) {
                riseTime = [self.datasource day];
            }
            setTime = event.eventTime;
        }
        if (riseTime && setTime) {
            [pairs addObject:@[riseTime,setTime]];
            riseTime = nil;
            setTime = nil;
        }
    }
    return [NSArray arrayWithArray:pairs];
}

-(void)drawRect:(CGRect)rect {
    DLog(@"*** Drawing chart!");
    CGFloat chartBottom = self.frame.size.height;
    
	CGContextRef context = UIGraphicsGetCurrentContext();
					  
	_tide = [self.datasource tideDataToChart];
    
    NSArray *intervalsForDay = [_tide intervalsFromDate:[self.datasource day] forHours:self.hoursToPlot];
    if ([intervalsForDay count] == 0) {
        // we're in a bad state. maybe activated on a new day before model has been updated?
        return;
    }
    
    NSTimeInterval baseSeconds = [((SDTideInterval*)intervalsForDay[0]).time timeIntervalSince1970];
    
    NSArray *sunEvents = [_tide sunriseSunsetEvents];
    NSArray *sunPairs = [self pairRiseAndSetEvents:sunEvents riseEventType:sunrise setEventType:sunset];

    NSArray *moonEvents = [_tide moonriseMoonsetEvents];
    NSArray *moonPairs = [self pairRiseAndSetEvents:moonEvents riseEventType:moonrise setEventType:moonset];
	
	// 480 x 320 = 24hrs x amplitude + some margin
	float min = [self findLowestTide:_tide];
	float max = [self findHighestTide:_tide];
	
	float ymin = min - 1;
	float ymax = max + 1;
	self.yratio =  _height / (ymax - ymin);
	self.yoffset = (_height + ymin * self.yratio) + (chartBottom - _height);
	//DLog(@"yoffset = %0.4f", self.yoffset);
	
	float xmin = 0;
	float xmax = MINUTES_PER_HOUR * _hoursToPlot;
	//DLog(@"Frame size is: %0.1f x %0.1f", self.frame.size.width, self.frame.size.height);
    self.xratio = self.frame.size.width / xmax;
	//DLog(@"xratio = %0.4f",self.xratio);
    
    // show daylight hours as light background
    CGContextSetRGBFillColor(context, 0.04, 0.27, 0.61, 1.0);
    
    for (NSArray *riseSet in sunPairs) {
        int sunriseMinutes = ([riseSet[0] timeIntervalSince1970] - baseSeconds) / SECONDS_PER_MINUTE;
        int sunsetMinutes = ([riseSet[1] timeIntervalSince1970] - baseSeconds) / SECONDS_PER_MINUTE;
        CGContextFillRect(context, CGRectMake(sunriseMinutes * self.xratio, 0, sunsetMinutes * self.xratio - sunriseMinutes * self.xratio, chartBottom));
    }
    
    CGContextSetRGBFillColor(context, 1, 1, 1, 0.2);
    for (NSArray *riseSet in moonPairs) {
        int moonriseMinutes = ([riseSet[0] timeIntervalSince1970] - baseSeconds) / SECONDS_PER_MINUTE;
        int moonsetMinutes = ([riseSet[1] timeIntervalSince1970] - baseSeconds) / SECONDS_PER_MINUTE;
        CGContextFillRect(context, CGRectMake(moonriseMinutes * self.xratio, 0, moonsetMinutes * self.xratio - moonriseMinutes * self.xratio, chartBottom));
    }
    
    // draws the tide level curve
    for (SDTideInterval *tidePoint in intervalsForDay) {
		int minute = ([[tidePoint time] timeIntervalSince1970] - baseSeconds) / SECONDS_PER_MINUTE;
        //DLog(@"Plotting interval: %@, min since midnight: %d",tidePoint.time, minute);
		if (minute == 0) {
			CGContextMoveToPoint(context, minute * self.xratio, self.yoffset - [tidePoint height] * self.yratio);
		} else {
			CGContextAddLineToPoint(context, (minute * self.xratio), self.yoffset - ([tidePoint height] * self.yratio));
		}
	}
	
    // closes the path so that it can be filled.
    int lastMinute = ([((SDTideInterval*)[intervalsForDay lastObject]).time timeIntervalSince1970] - baseSeconds) / SECONDS_PER_MINUTE;
	CGContextAddLineToPoint(context, lastMinute*self.xratio, chartBottom);
	CGContextAddLineToPoint(context, 0, chartBottom);
	
    // fill in the tide level curve
	CGContextSetRGBFillColor(context, 0.0, 1.0, 1.0, 0.7);
	CGContextFillPath(context);
    
    // Drawing with a white stroke color
	CGContextSetRGBStrokeColor(context, 1.0, 1.0, 1.0, 1.0);
    
    // draws the zero height line
    CGContextSetLineWidth(context,2.0);
	CGContextMoveToPoint(context, xmin, _yoffset);
	CGContextAddLineToPoint(context, xmax * _xratio, _yoffset);
	CGContextStrokePath(context);

	NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
	[formatter setDateStyle:NSDateFormatterFullStyle];
	self.dateLabel.text = [formatter stringFromDate:[self.datasource day]];
    DLog(@"*** Finished drawing chart. xratio=%f", self.xratio);
}

-(void)hideTideDetails
{
	self.valueLabel.text = @"";
}
                                           
- (NSDate*)midnight {
    return [self midnight: [NSDate date]];
}

- (NSDate*)midnight: (NSDate*)date {
	NSCalendar *gregorian = [NSCalendar currentCalendar];
	unsigned unitflags = NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay;
	NSDateComponents *components = [gregorian components: unitflags fromDate: date];
	
	return [gregorian dateFromComponents:components];
}	

- (float)findLowestTide:(SDTide *)tide {	
	NSSortDescriptor *ascDescriptor = [[NSSortDescriptor alloc] initWithKey:@"height" ascending:YES];
	NSArray *descriptors = @[ascDescriptor];
	NSArray *ascResult = [[tide allIntervals] sortedArrayUsingDescriptors:descriptors];
	return [(SDTideInterval*)ascResult[0] height];
}

- (float)findHighestTide:(SDTide *)tide {
	NSSortDescriptor *descDescriptor = [[NSSortDescriptor alloc] initWithKey:@"height" ascending:NO];
	NSArray *descriptors = @[descDescriptor];
	NSArray *descResult = [[tide allIntervals] sortedArrayUsingDescriptors:descriptors];
	return [(SDTideInterval*)descResult[0] height];
}

@end

File: Classes/Views/ChartView.swift
//
//  File.swift
//  ShralpTide2
//
//  Created by Michael Parlee on 10/3/16.
//
//

import Foundation
import UIKit

@objc class ChartView: UIView {
    
    internal static let MinutesPerHour = 60
    internal static let SecondsPerMinute = 60

    fileprivate let dateFormatter = DateFormatter()
    
    internal var sunriseIcon:UIImage?
    internal var sunsetIcon:UIImage?
    internal var moonriseIcon:UIImage?
    internal var moonsetIcon:UIImage?
    
    @objc var hoursToPlot:Int = 24
    @objc var height:Int = 234
    
    @objc var showZero = true
    
    @objc var page:Int = 0
    
    var xratio: CGFloat = 0
    var yratio: CGFloat = 0
    var yoffset: CGFloat = 0
    
    @IBOutlet var datasource: ChartViewDatasource!
    @IBOutlet var cursorView: UIView!
    @IBOutlet var dateLabel: UILabel?
    @IBOutlet var valueLabel: UILabel!
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        dateFormatter.dateStyle = .full
    }
    
    fileprivate func endTime() throws -> Date  {
        return Date(timeIntervalSince1970: datasource.day.timeIntervalSince1970 + Double(hoursToPlot) * Double(ChartView.MinutesPerHour * ChartView.SecondsPerMinute))
    }
    
    fileprivate func pairRiseAndSetEvents(_ events:[SDTideEvent], riseEventType:SDTideState, setEventType:SDTideState) throws -> Array<(Date,Date)> {
        var pairs:Array<(Date,Date)> = [(Date,Date)]()
        var riseTime:Date!
        var setTime:Date!
        for event:SDTideEvent in events {
            if event.eventType == riseEventType {
                riseTime = event.eventTime
                if event === events.last {
                    setTime = try endTime()
                }
            }
            if event.eventType == setEventType {
                if events.index(of:event) == 0 {
                    riseTime = datasource.day
                }
                setTime = event.eventTime
            }
            if riseTime != nil && setTime != nil {
                pairs.append((riseTime,setTime))
                riseTime = nil
                setTime = nil
            }
        }
        let immutablePairs = pairs
        return immutablePairs
    }
    
    func midnight() -> Date {
        return midnight(Date())
    }
    
    func midnight(_ date:Date) -> Date {
        let date = Date()
        let calendar = Calendar(identifier: .gregorian)
        return calendar.startOfDay(for: date)
    }
    
    fileprivate func findLowestTideValue(_ tide:SDTide) -> CGFloat {
        return CGFloat(tide.allIntervals.sorted( by: { $0.height < $1.height } )[0].height)
    }
    
    fileprivate func findHighestTideValue(_ tide:SDTide) -> CGFloat {
        return CGFloat(tide.allIntervals.sorted( by: { $0.height > $1.height } )[0].height)
    }
    
    override func draw(_ rect:CGRect) {
        let chartBottom:CGFloat = frame.size.height
        
        guard let tide = datasource.tideDataToChart else {
            print("No tide data found.")
            return
        }
        
        let intervalsForDay:[SDTideInterval] = tide.intervals(from: datasource.day, forHours: hoursToPlot)
        
        guard intervalsForDay.count > 0 else {
            return
        }
        
        do {
            let baseSeconds:TimeInterval = intervalsForDay[0].time.timeIntervalSince1970
            
            let sunEvents:[SDTideEvent] = tide.sunriseSunsetEvents
            
            let sunPairs:Array<(Date,Date)> = try pairRiseAndSetEvents(sunEvents, riseEventType: .sunrise, setEventType: .sunset)
            
            let moonEvents:[SDTideEvent] = tide.moonriseMoonsetEvents
            
            let moonPairs:Array<(Date,Date)> = try pairRiseAndSetEvents(moonEvents, riseEventType: .moonrise, setEventType: .moonset)
            
            let min:CGFloat = findLowestTideValue(tide)
            let max:CGFloat = findHighestTideValue(tide)
            
            let ymin:CGFloat = min - 1
            let ymax:CGFloat = max + 1
            
            yratio = CGFloat(height) / (ymax - ymin)
            yoffset = (CGFloat(height) + ymin * self.yratio) + (chartBottom - CGFloat(height))
            
            let xmin:Int = 0
            let xmax:Int = ChartView.MinutesPerHour * hoursToPlot
            
            xratio = CGFloat(frame.size.width) / CGFloat(xmax)
            
            guard let context = UIGraphicsGetCurrentContext() else {
                print("Failed to get graphics context.")
                return
            }
            
            // show daylight hours as light background
            let dayColor = UIColor(red:0.04, green:0.27, blue:0.61, alpha:1)
            dayColor.setFill()
            dayColor.setStroke()
            
            for (rise,set) in sunPairs {
                let sunriseMinutes = Int(rise.timeIntervalSince1970 - baseSeconds) / ChartView.SecondsPerMinute
                let sunsetMinutes = Int(set.timeIntervalSince1970 - baseSeconds) / ChartView.SecondsPerMinute
                context.addPath(CGPath(rect:CGRect(x:CGFloat(sunriseMinutes) * xratio, y:0, width:CGFloat(sunsetMinutes) * xratio - CGFloat(sunriseMinutes) * xratio, height:chartBottom), transform:nil))
                context.fillPath()
            }
            
            let moonColor = UIColor(red:1, green:1, blue: 1, alpha:0.2)
            moonColor.setStroke()
            moonColor.setFill()
            for (rise,set) in moonPairs {
                let moonriseMinutes = Int(rise.timeIntervalSince1970 - baseSeconds) / ChartView.SecondsPerMinute
                let moonsetMinutes = Int(set.timeIntervalSince1970 - baseSeconds) / ChartView.SecondsPerMinute
                context.addPath(CGPath(rect:CGRect(x:CGFloat(moonriseMinutes) * xratio, y:0, width:CGFloat(moonsetMinutes) * xratio - CGFloat(moonriseMinutes) * xratio, height:chartBottom), transform:nil))
                context.fillPath()
            }
            
            //draws the tide level curve
            let tidePath = UIBezierPath()
            for tidePoint:SDTideInterval in intervalsForDay {
                let minute:Int = Int(tidePoint.time.timeIntervalSince1970 - baseSeconds) / ChartView.SecondsPerMinute
                if minute == 0 {
                    tidePath.move(to: CGPoint(x:CGFloat(minute) * xratio, y:yoffset - CGFloat(tidePoint.height) * yratio))
                } else {
                    tidePath.addLine(to: CGPoint(x:CGFloat(minute) * xratio, y:yoffset - CGFloat(tidePoint.height) * yratio))
                }
            }
            
            // closes the path so it can be filled
            let lastMinute:Int = Int(intervalsForDay.last!.time.timeIntervalSince1970 - baseSeconds) / ChartView.SecondsPerMinute
            tidePath.addLine(to:CGPoint(x:CGFloat(lastMinute) * xratio, y: chartBottom))
            tidePath.addLine(to:CGPoint(x:0, y:chartBottom))
            
            // fill in the tide level curve
            let tideColor = UIColor(red:0, green:1, blue: 1, alpha:0.7)
            tideColor.setStroke()
            tideColor.setFill()

            context.addPath(tidePath.cgPath)
            context.fillPath()
            
            // drawing with a white stroke color
            let axisColor = UIColor(red:1, green:1, blue:1, alpha:1)
            axisColor.setStroke()
            
            if showZero {
                // draws the zero height line
                let baseLinePath = UIBezierPath()
                context.setLineWidth(2)
                baseLinePath.move(to:CGPoint(x:CGFloat(xmin),y:CGFloat(yoffset)))
                baseLinePath.addLine(to:CGPoint(x:CGFloat(xmax) * CGFloat(xratio), y:CGFloat(yoffset)))
                
                context.addPath(baseLinePath.cgPath)
                context.strokePath()
            }
            
            dateLabel?.text = dateFormatter.string(from: datasource.day)
        } catch {
            print("Error drawing chart")
            return
        }
    }
}
